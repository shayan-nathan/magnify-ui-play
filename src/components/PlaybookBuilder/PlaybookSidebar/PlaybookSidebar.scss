@import './ActionSource/ActionSource.scss';
@import './SidebarSearch/SidebarSearch.scss';

.playbook-sidebar {
  overflow-y: auto;

  .c-sider__header {
    font-weight: 600;
    font-size: 24px;
    line-height: 24px;
    color: $color-primary;
    border-width: 0;
  }

  .action-group {
    margin-bottom: 24px;
  }

  .ant-menu {
    color: $color-secondary-dark-blue;
  }

  .submenu-icon {
    background-color: rgba(0, 0, 0, 0.05);
    border-radius: 10px;
    position: absolute;
    width: 24px;
    height: 20px;
    font-size: 14px;
    right: 0;
    display: flex;
    justify-content: center;
    align-items: center;

    &.icon-reverse {
      display: none;
    }
  }

  .ant-menu-sub {
    &.ant-menu-inline {
      background-color: transparent;
    }
  }

  ant-menu-inline {
    > .ant-menu-submenu {
      > .ant-menu-submenu-title {
        height: 28px;
        line-height: 28px;
      }
    }
  }

  .ant-menu-inline .ant-menu-submenu-title {
    margin-top: 0;
    margin-bottom: 10px;
  }

  .ant-menu-title-content {
    border-width: 0;
    margin-left: 8px;
  }

  .ant-menu-submenu {
    .ant-menu-submenu-title {
      &:hover {
        color: $color-font-main;

        .submenu-icon {
          display: none;

          &.icon-reverse {
            background-color: #d6d7d8;
            display: flex;
            transform: rotate(180deg);
          }
        }
      }
    }

    &-open {
      .ant-menu-submenu-title {
        &:hover {
          .submenu-icon {
            &.icon-reverse {
              display: flex;
              transform: rotate(0deg);
            }
          }
        }

        .submenu-icon {
          display: none;

          &.icon-reverse {
            display: flex;
          }
        }
      }
    }
  }

  .action-group:nth-of-type(2) {
    .action-source:not(:last-child) {
      margin-bottom: 4px;
    }
  }

  &::-webkit-scrollbar {
    width: 2px;
  }

  @include thin-scrollbar;
}
