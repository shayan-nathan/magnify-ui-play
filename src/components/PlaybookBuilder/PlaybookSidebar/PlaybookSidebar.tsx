import { Menu, SubMenuProps } from 'antd'
import Sider from 'antd/lib/layout/Sider'
import { FC, useEffect, useMemo, useState } from 'react'

import { LogoService } from 'services/Utils/logo'
import { Heading } from 'components/common'
import { IconArrowTop, IconSearch } from 'components/common/Icons/Icons'
import { Integration } from 'models/Integration.model'
import { Action, PlaybookActionsType, Platform } from '../PlaybookBuilderTypes'
import { playbookBuilderOptions } from './Settings/PlaybookBuilder.settings'
import { ActionSource } from './ActionSource/ActionSource'
import { SidebarSearch } from './SidebarSearch/SidebarSearch'
import useWindowSize from 'hooks/useWindowSize'

interface PlaybookSidebarProps {
  platforms: Integration[]
}

export const PlaybookSidebar: FC<PlaybookSidebarProps> = () => {
  const { SubMenu } = Menu
  const [filteredActions, setFilteredActions] = useState<PlaybookActionsType>(playbookBuilderOptions)
  const { height } = useWindowSize()
  const [sidebarHeight, setSidebarHeight] = useState(height)
  // TODO: filter the platform based on the connected apps
  const expandMenuIcon = (props: SubMenuProps | any) => {
    const noOfChildren = (props.children as Array<Element>).length
    let menuIcon = <span className='submenu-icon'>{noOfChildren}</span>
    return menuIcon
  }
  const displayActions = (category: string, categorySources: Action[] | Platform[]): JSX.Element[] | JSX.Element => {
    if (category !== 'actions') {
      return (categorySources as Action[]).map((source) => {
        return (
          <ActionSource
            key={source.name}
            action={{ ...source, iconName: source.type, payload: {}, description: source.name }}></ActionSource>
        )
      })
    }
    const platforms = categorySources as Platform[]
    const platformElements = platforms.map((platform) => {
      const platformActions = platform.actions
      const actions = platformActions.map((platform_source) => {
        return (
          <ActionSource
            key={platform_source.name}
            action={{
              ...platform_source,
              platform: platform.name,
              iconName: platform_source.type,
              payload: {},
              description: platform_source.name,
            }}></ActionSource>
        )
      })
      return (
        <SubMenu
          key={platform.name}
          title={platform.name}
          icon={
            <>
              <img src={`${LogoService.getIcon(platform.name)}`} alt={platform.name} />

              <span className='submenu-icon icon-reverse'>
                <IconArrowTop />
              </span>
            </>
          }
          data-testid='submenu-platform'>
          {actions}
        </SubMenu>
      )
    })

    const platformsName = platforms.map((platform) => platform.name)

    return (
      <Menu
        style={{ width: '100%' }}
        inlineIndent={0}
        expandIcon={expandMenuIcon}
        mode='inline'
        data-testid='action-menu'
        defaultOpenKeys={platformsName}>
        {platformElements}
      </Menu>
    )
  }

  const renderElements = (actions: PlaybookActionsType) => {
    const elements = []
    for (const category in actions) {
      const categorySources = (actions as unknown as { [key: string]: Action[] | Platform[] })[category]
      const categoryElements = (
        <div key={category} className='action-group'>
          <Heading level='5' variant='5'>
            {category.toUpperCase()}
          </Heading>

          {displayActions(category, categorySources)}
        </div>
      )
      elements.push(categoryElements)
    }
    return elements
  }

  const memoizedElements = useMemo(() => {
    return renderElements(filteredActions)
  }, [filteredActions])

  useEffect(() => {
    setSidebarHeight(height)
  }, [height])

  return (
    <div>
      <Sider
        className='c-sider playbook-sidebar'
        width={265}
        data-testid='playbook-sidebar'
        style={{ height: `${sidebarHeight}px` }}>
        {/* <span className='c-sider__header' data-testid='playbook-header'>
        Playbook builder
      </span> */}
        {/* EN-477 */}
        <span className='c-sider__header' data-testid='playbook-header'>
          Journey builder
        </span>
        <SidebarSearch prefix={<IconSearch />} onChangeState={setFilteredActions} />

        {memoizedElements}
      </Sider>
    </div>
  )
}
