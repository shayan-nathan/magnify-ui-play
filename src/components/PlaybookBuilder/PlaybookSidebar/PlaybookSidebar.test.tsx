import { render, screen } from '@testing-library/react'
import userEvent from '@testing-library/user-event'

import { PlaybookSidebar } from './PlaybookSidebar'

describe('<PlaybookSidebar /> component', () => {
  it('renders correctly all the elements', async () => {
    render(<PlaybookSidebar platforms={[]} />)
    const playbookSidebar = await screen.findByTestId('playbook-sidebar')
    expect(playbookSidebar).toMatchSnapshot()

    const sidebarHeader = await screen.findByTestId('playbook-header')
    expect(sidebarHeader).toBeTruthy()

    const searchElement: HTMLInputElement = await screen.findByTestId('search')
    expect(searchElement).toBeTruthy()

    const actionMenuElement = await screen.findByTestId('action-menu')
    expect(actionMenuElement).toBeTruthy()

    const actionSources = await screen.findAllByTestId('action-source')
    expect(actionSources.length).toBe(11)
    expect(actionSources).toBeTruthy()

    const groupHeadigs = await screen.findAllByTestId('heading')
    expect(groupHeadigs).toBeTruthy()
    expect(groupHeadigs.length).toBe(3)
    expect(groupHeadigs[0].textContent).toBe('TARGETING')
    expect(groupHeadigs[1].textContent).toBe('RULES')
    expect(groupHeadigs[2].textContent).toBe('ACTIONS')
  })

  it('behaves correctly using the search field', async () => {
    render(<PlaybookSidebar platforms={[]} />)
    const searchElement: HTMLInputElement = await screen.findByTestId('search')
    expect(searchElement).toBeTruthy()
    // initially the field is empty
    expect(searchElement).toHaveValue('')
    const initialActionSources = await screen.findAllByTestId('action-source')
    expect(initialActionSources.length).toBe(11)
    // simulate the search for "time del"
    userEvent.type(searchElement, 'time del')
    expect(searchElement).toHaveValue('time del')

    const actionSources = await screen.findAllByTestId('action-source')
    expect(actionSources.length).toBe(1)
    expect(actionSources[0].textContent).toBe('Time delay')

    const submenuPlatforms = await screen.findAllByTestId('submenu-platform')
    expect(submenuPlatforms.length).toBe(2)
    expect(submenuPlatforms[0].textContent).toContain('Salesforce')

    // simulate clearing the whole input
    userEvent.clear(searchElement)
    expect(searchElement).toHaveValue('')
    const actionsAfterClearingTheSearch = await screen.findAllByTestId('action-source')
    expect(actionsAfterClearingTheSearch.length).toBe(11)
  })
})
