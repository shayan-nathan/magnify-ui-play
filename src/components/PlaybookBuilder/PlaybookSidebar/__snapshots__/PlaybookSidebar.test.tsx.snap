// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<PlaybookSidebar /> component renders correctly all the elements 1`] = `
<aside
  class="ant-layout-sider ant-layout-sider-dark c-sider playbook-sidebar"
  data-testid="playbook-sidebar"
  style="flex: 0 0 265px; max-width: 265px; min-width: 265px; width: 265px;"
>
  <div
    class="ant-layout-sider-children"
  >
    <span
      class="c-sider__header"
      data-testid="playbook-header"
    >
      Playbook builder
    </span>
    <span
      class="ant-input-affix-wrapper"
    >
      <span
        class="ant-input-prefix"
      >
        <svg>
          search.svg
        </svg>
      </span>
      <input
        class="ant-input"
        data-testid="search"
        placeholder="search"
        type="text"
        value=""
      />
    </span>
    <div
      class="action-group"
    >
      <h5
        class="c-heading c-heading--5 undefined"
        data-testid="heading"
      >
        TARGETING
      </h5>
      <div
        class="action-source"
        data-testid="action-source"
        draggable="true"
      >
        <div
          class="action-btn--box"
          data-testid="source-box"
        >
          <button
            class="ant-btn c-targeting c-targeting--circle"
            data-testid="source-btn"
            type="button"
          >
            <img
              alt="Segments"
              data-testid="source-image"
              draggable="false"
              src="segments.svg"
            />
          </button>
        </div>
        <div
          class="source-name"
          data-testid="source-name"
        >
          <span>
            Segments
          </span>
        </div>
      </div>
    </div>
    <div
      class="action-group"
    >
      <h5
        class="c-heading c-heading--5 undefined"
        data-testid="heading"
      >
        RULES
      </h5>
      <div
        class="action-source"
        data-testid="action-source"
        draggable="true"
      >
        <div
          class="action-btn--box"
          data-testid="source-box"
        >
          <button
            class="ant-btn c-targeting c-targeting--rhomb"
            data-testid="source-btn"
            type="button"
          >
            <img
              alt="If / Else"
              data-testid="source-image"
              draggable="false"
              src="if_else.svg"
            />
          </button>
        </div>
        <div
          class="source-name"
          data-testid="source-name"
        >
          <span>
            If / Else
          </span>
        </div>
      </div>
      <div
        class="action-source"
        data-testid="action-source"
        draggable="true"
      >
        <div
          class="action-btn--box"
          data-testid="source-box"
        >
          <button
            class="ant-btn c-targeting c-targeting--rhomb"
            data-testid="source-btn"
            type="button"
          >
            <img
              alt="Wait for trigger"
              data-testid="source-image"
              draggable="false"
              src="wait_for_trigger.svg"
            />
          </button>
        </div>
        <div
          class="source-name"
          data-testid="source-name"
        >
          <span>
            Wait for trigger
          </span>
        </div>
      </div>
      <div
        class="action-source"
        data-testid="action-source"
        draggable="true"
      >
        <div
          class="action-btn--box"
          data-testid="source-box"
        >
          <button
            class="ant-btn c-targeting c-targeting--rhomb"
            data-testid="source-btn"
            type="button"
          >
            <img
              alt="Time delay"
              data-testid="source-image"
              draggable="false"
              src="time_delay.svg"
            />
          </button>
        </div>
        <div
          class="source-name"
          data-testid="source-name"
        >
          <span>
            Time delay
          </span>
        </div>
      </div>
      <div
        class="action-source"
        data-testid="action-source"
        draggable="true"
      >
        <div
          class="action-btn--box"
          data-testid="source-box"
        >
          <button
            class="ant-btn c-targeting c-targeting--rhomb"
            data-testid="source-btn"
            type="button"
          >
            <img
              alt="Loop"
              data-testid="source-image"
              draggable="false"
              src="loop.svg"
            />
          </button>
        </div>
        <div
          class="source-name"
          data-testid="source-name"
        >
          <span>
            Loop
          </span>
        </div>
      </div>
    </div>
    <div
      class="action-group"
    >
      <h5
        class="c-heading c-heading--5 undefined"
        data-testid="heading"
      >
        ACTIONS
      </h5>
      <ul
        class="ant-menu ant-menu-root ant-menu-inline ant-menu-light"
        data-menu-list="true"
        data-testid="action-menu"
        role="menu"
        style="width: 100%;"
        tabindex="0"
      >
        <li
          class="ant-menu-submenu ant-menu-submenu-inline"
          data-testid="submenu-platform"
          role="none"
        >
          <div
            aria-controls="rc-menu-uuid-test-Salesforce-popup"
            aria-expanded="false"
            aria-haspopup="true"
            class="ant-menu-submenu-title"
            data-menu-id="rc-menu-uuid-test-Salesforce"
            role="menuitem"
            style="padding-left: 0px;"
            tabindex="-1"
          >
            <svg>
              salesforce.svg
            </svg>
            <span
              class="submenu-icon icon-reverse"
            >
              <svg>
                arrow_top.svg
              </svg>
            </span>
            <span
              class="ant-menu-title-content"
            >
              Salesforce
            </span>
            <span
              class="submenu-icon"
            >
              3
            </span>
          </div>
        </li>
        <li
          class="ant-menu-submenu ant-menu-submenu-inline"
          data-testid="submenu-platform"
          role="none"
        >
          <div
            aria-controls="rc-menu-uuid-test-Marketo-popup"
            aria-expanded="false"
            aria-haspopup="true"
            class="ant-menu-submenu-title"
            data-menu-id="rc-menu-uuid-test-Marketo"
            role="menuitem"
            style="padding-left: 0px;"
            tabindex="-1"
          >
            <svg>
              marketo.svg
            </svg>
            <span
              class="submenu-icon icon-reverse"
            >
              <svg>
                arrow_top.svg
              </svg>
            </span>
            <span
              class="ant-menu-title-content"
            >
              Marketo
            </span>
            <span
              class="submenu-icon"
            >
              3
            </span>
          </div>
        </li>
      </ul>
      <div
        aria-hidden="true"
        style="display: none;"
      >
        <div
          class="action-source"
          data-testid="action-source"
          draggable="true"
        >
          <div
            class="action-btn--box"
            data-testid="source-box"
          >
            <button
              class="ant-btn c-targeting c-targeting--square"
              data-testid="source-btn"
              type="button"
            >
              <img
                alt="Update lead"
                data-testid="source-image"
                draggable="false"
                src="update_lead.svg"
              />
            </button>
          </div>
          <div
            class="source-name"
            data-testid="source-name"
          >
            <span>
              Update lead
            </span>
          </div>
        </div>
        <div
          class="action-source"
          data-testid="action-source"
          draggable="true"
        >
          <div
            class="action-btn--box"
            data-testid="source-box"
          >
            <button
              class="ant-btn c-targeting c-targeting--square"
              data-testid="source-btn"
              type="button"
            >
              <img
                alt="Send email"
                data-testid="source-image"
                draggable="false"
                src="send_email.svg"
              />
            </button>
          </div>
          <div
            class="source-name"
            data-testid="source-name"
          >
            <span>
              Send email
            </span>
          </div>
        </div>
        <div
          class="action-source"
          data-testid="action-source"
          draggable="true"
        >
          <div
            class="action-btn--box"
            data-testid="source-box"
          >
            <button
              class="ant-btn c-targeting c-targeting--square"
              data-testid="source-btn"
              type="button"
            >
              <img
                alt="Create opportunity"
                data-testid="source-image"
                draggable="false"
                src="create_opportunity.svg"
              />
            </button>
          </div>
          <div
            class="source-name"
            data-testid="source-name"
          >
            <span>
              Create opportunity
            </span>
          </div>
        </div>
        <div
          class="action-source"
          data-testid="action-source"
          draggable="true"
        >
          <div
            class="action-btn--box"
            data-testid="source-box"
          >
            <button
              class="ant-btn c-targeting c-targeting--square"
              data-testid="source-btn"
              type="button"
            >
              <img
                alt="Update lead"
                data-testid="source-image"
                draggable="false"
                src="update_lead.svg"
              />
            </button>
          </div>
          <div
            class="source-name"
            data-testid="source-name"
          >
            <span>
              Update lead
            </span>
          </div>
        </div>
        <div
          class="action-source"
          data-testid="action-source"
          draggable="true"
        >
          <div
            class="action-btn--box"
            data-testid="source-box"
          >
            <button
              class="ant-btn c-targeting c-targeting--square"
              data-testid="source-btn"
              type="button"
            >
              <img
                alt="Send email"
                data-testid="source-image"
                draggable="false"
                src="send_email.svg"
              />
            </button>
          </div>
          <div
            class="source-name"
            data-testid="source-name"
          >
            <span>
              Send email
            </span>
          </div>
        </div>
        <div
          class="action-source"
          data-testid="action-source"
          draggable="true"
        >
          <div
            class="action-btn--box"
            data-testid="source-box"
          >
            <button
              class="ant-btn c-targeting c-targeting--square"
              data-testid="source-btn"
              type="button"
            >
              <img
                alt="Create opportunity"
                data-testid="source-image"
                draggable="false"
                src="create_opportunity.svg"
              />
            </button>
          </div>
          <div
            class="source-name"
            data-testid="source-name"
          >
            <span>
              Create opportunity
            </span>
          </div>
        </div>
      </div>
    </div>
  </div>
</aside>
`;
